cmd/plugins/topology-aware/policy/resources.go

```diff
@@ -997,12 +1013,81 @@ func (cs *supply) AllocatableSharedCPU(quiet ...bool) int {
 			shared = pShared
 		}
 	}
+
 	if verbose {
 		log.Debug("%s: ancestor-adjusted free shared CPU: %dm", cs.node.Name(), shared)
 	}
+
+	// If there are BestEffort containers in the node or any of its children
+	// we need to account for them an extra milliCPU worth of shared capacity.
+	//
+	// TODO(klihub): We might need to try speeding this up if it gets too slow
+	// on nodes with a lot of CPUs and topology zones as the number of containers
+	// grows. hasZeroCpuReqContainer uses foreachContainer, which in turn loops
+	// through each resource grant calling a callback if the grant is assigned to
+	// a particular node. foreachContainer can stop early based on the callback's
+	// return value, but {Breadth,Depth}First can not ATM. Obvious optimizations
+	// would be to 1) allow {Breadth,Depth}First to stop early if possible, and
+	// 2) store grants per assigned node.
+	hasBestEffort := false
+	cs.node.BreadthFirst(func(n Node) {
+		if cs.node.Policy().hasZeroCpuReqContainer(n) {
+			hasBestEffort = true
+		}
+	})
+	if hasBestEffort {
+		shared--
+	}
+
 	return shared
 }
 
+// SlicableCPUs calculates the shared cpuset we can slice exclusive CPUs off of.
+func (cs *supply) SlicableCPUs() (cpuset.CPUSet, error) {
+	var (
+		slicable = cpuset.New()
+		errs     []error
+	)
+
+	// We need to avoid slicing any shared pool in child nodes below its
+	// current allocation. To do so we go through the subtree collecting
+	// CPUs from shared pools for allocation but leaving off enough full
+	// CPUs in each for their current BestEffort and Burstable QoS class
+	// shared allocations.
+
+	cs.node.DepthFirst(func(n Node) {
+		if n.IsSameNode(cs.node) && !n.IsLeafNode() {
+			return
+		}
+
+		ns := n.FreeSupply()
+		if ns == nil {
+			return
+		}
+
+		cpus := ns.SharableCPUs()
+		free := ns.AllocatableSharedCPU(true) / 1000
+
+		// TODO(klihub): We should ideally take also into account the CPU
+		// priority preference of any ongoing allocation, trying to slice
+		// CPUs with a matching preference. We don't do that ATM.
+
+		cset, err := cs.takeCPUs(&cpus, nil, free, nonePrio)
+		if err != nil {
+			errs = append(errs, err)
+			return
+		}
+
+		slicable = slicable.Union(cset)
```

Comment:
Not fully sure, so this is more of a question rather than suggestion:

Assume a parent node P has two child nodes, C1 and C2. C1 has shared CPUs cpu10 and cpu11, and C2 shared CPUs cpu20 and cpu21. Both C1 and C2 host one BestEffort container each. Parent node has no containers.

Does this mean that parent node P has shared CPUs cpu10, cpu11, cpu20, and cpu21? And that depth-first union could first go to C1, make slicable {cpu10}, then to C2 and make slicable {cpu10}.Union({cpu20}), and finally handle parent P, and make slicable {cpu10,cpu21}.Union({cpu10,cpu11,cpu20})?